students/john_navitsky/session06/mailroom.py

```diff
@@ -168,13 +182,52 @@ def print_thank_you(donor_id,hint="wonderful",dest=sys.stdout):
     print_lines(2,dest)
     print("-"*80,file=dest)
 
-    print(letter.format(hint,**donors[donor_id]),file=dest)
+    print(letter.format(hint,**current_donor),file=dest)
 
     print("-"*80,file=dest)
     print_lines(2,dest)
 
 
-def thank_you_entry():
+def match_donor(current_donor, donors):
+        """ See if record matches existing record, if so, return full record. """
+        for donor_id in donors:
+            existing_donor=donors[donor_id]
+            if existing_donor["full_name"].lower() == current_donor["full_name"].lower():
+                # if the owner exists, pull in the complete donor record
+                current_donor=existing_donor
+        return current_donor
+
+
+def update_donor(current_donor, donors, new_donation):
+        """ Update the donor database with new donation or donor record as needed. """
+
+        # match to existing record, if exists
+        current_donor = match_donor(current_donor, donors)
+
+        # datetime.datetime.strptime(<iso date>,'%Y-%m-%dZ')
+        today = datetime.datetime.utcnow().date().isoformat() + "Z"
+
+        # add donation to an existing donor
+        if "donor_id" in current_donor.keys():
+            donor_id=current_donor["donor_id"]
+            donors[donor_id]["donations"].append({ "amount": new_donation, "date": today })
+            hint="returning"
+        else:
+            # datetime.datetime.strptime(<iso time>,'%Y-%m-%dT%H:%M:%S.%fZ')
+            now = datetime.datetime.utcnow().isoformat() + "Z"
+            # id is a hash of the full donor name plus high resolution time
```

Comment:
if you are trying to simply get a unique id --maybe use uuid:

import uuid.

A couple other notes:

What function does an id serve here anyway?

and if you have an id in a donor record, why not simiply call it "id" -- if it's part of a donor record, we know it's a donor id.

Reply:
Yes, I found UUID and went back and forth about which approach.  In the end I think I picked a hash mostly because it was a bit shorter and should serve same function.

I figured having the id as the key and in the record might raise an eyebrow.  In general I found it was a convenience.

Good point on the redundancy.